File: Packages/Yage/Tests/YageTests/CollisionsTests.swift
import XCTest

@testable import Yage

class CollisionsTests: XCTestCase {
    private lazy var testEnv: World = {
        World(name: "test", bounds: testBounds)
    }()
    private let testBounds = CGRect(x: 0, y: 0, width: 100, height: 100)

    func testMultipleCollisionsCanBeDetected() {
        let main = Entity(
            species: .agent,
            id: "main",
            frame: CGRect(x: 0, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let others: [Entity] = (0 ..< 12).map { (count: Int) in
            Entity(
                species: .agent,
                id: "other-\(count)",
                frame: CGRect(
                    x: 0.1 * Double(count),
                    y: 0, width: 1, height: 1
                ),
                in: testEnv
            )
        }
        let collisions = main.collisions(with: others)
        XCTAssertEqual(collisions.count, 11)

        let moreCollisions = main.collisions(with: [main] + others)
        XCTAssertEqual(moreCollisions.count, 11)
    }

    func testDistantEntitiesDoNotCollide() {
        let entity1 = Entity(
            species: .agent,
            id: "entity1",
            frame: CGRect(x: 0, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let entity2 = Entity(
            species: .agent,
            id: "entity2",
            frame: CGRect(x: 2, y: 2, width: 1, height: 1),
            in: testEnv
        )
        let collision = entity1.collision(with: entity2)
        XCTAssertNil(collision)
    }

    func testEqualEntitiesDoCollide() {
        let entity = Entity(species: .agent, id: "someEntity", frame: .zero, in: testEnv)
        let collision = entity.collision(with: entity)
        XCTAssertNotNil(collision)
    }

    func testEntitiesWithSameFrameCollide() {
        let entity1 = Entity(species: .agent, id: "entity1", frame: .zero, in: testEnv)
        let entity2 = Entity(species: .agent, id: "entity2", frame: .zero, in: testEnv)
        let collision = entity1.collision(with: entity2)
        XCTAssertNotNil(collision)
    }

    func testEntitiesSharingOneCornerCollide() {
        let entity1 = Entity(
            species: .agent,
            id: "entity1",
            frame: CGRect(x: 0, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let entity2 = Entity(
            species: .agent,
            id: "entity2",
            frame: CGRect(x: 1, y: 1, width: 1, height: 1),
            in: testEnv
        )
        let collision = entity1.collision(with: entity2)
        let expectedIntersection = CGRect(x: 1, y: 1, width: 0, height: 0)
        XCTAssertNotNil(collision)
        XCTAssertEqual(collision?.intersection ?? .zero, expectedIntersection)
    }

    func testEntitiesSharingOneSideCollide() {
        let entity1 = Entity(
            species: .agent,
            id: "entity1",
            frame: CGRect(x: 0, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let entity2 = Entity(
            species: .agent,
            id: "entity2",
            frame: CGRect(x: 1, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let collision = entity1.collision(with: entity2)
        let expectedIntersection = CGRect(x: 1, y: 0, width: 0, height: 1)
        XCTAssertNotNil(collision)
        XCTAssertEqual(collision?.intersection ?? .zero, expectedIntersection)
    }

    func testOverlappingEntitiesCollide() {
        let entity1 = Entity(
            species: .agent,
            id: "entity1",
            frame: CGRect(x: 0, y: 0, width: 2, height: 2),
            in: testEnv
        )
        let entity2 = Entity(
            species: .agent,
            id: "entity2",
            frame: CGRect(x: 1, y: 1, width: 2, height: 2),
            in: testEnv
        )
        let collision = entity1.collision(with: entity2)
        let expectedIntersection = CGRect(x: 1, y: 1, width: 1, height: 1)
        XCTAssertNotNil(collision)
        XCTAssertEqual(collision?.intersection ?? .zero, expectedIntersection)
    }

    func testCollisionWithEphimeralEntitiesAreProperlyMarked() {
        let entity1 = Entity(species: .agent, id: "entity1", frame: .zero, in: testEnv)
        let entity2 = Entity(species: .agent, id: "entity2", frame: .zero, in: testEnv)
        entity2.isEphemeral = true
        let collision = entity1.collision(with: entity2)
        XCTAssertTrue(collision?.isEphemeral ?? false)
    }

    func testSidesDetectedWhenSameHeightObjectPerfectlyCollidesOnRight() {
        let entity1 = Entity(
            species: .agent,
            id: "entity1",
            frame: CGRect(x: 0, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let entity2 = Entity(
            species: .agent,
            id: "entity2",
            frame: CGRect(x: 0.9, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let collision = entity1.collision(with: entity2)
        let sides = collision?.sides() ?? []
        XCTAssertEqual(sides, [.top, .right, .bottom])
    }

    func testSidesDetectedWhenSmallerCollidesOnRight() {
        let entity1 = Entity(
            species: .agent,
            id: "entity1",
            frame: CGRect(x: 0, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let entity2 = Entity(
            species: .agent,
            id: "entity2",
            frame: CGRect(x: 0.9, y: 0.25, width: 1, height: 0.5),
            in: testEnv
        )
        let collision = entity1.collision(with: entity2)
        let sides = collision?.sides() ?? []
        XCTAssertEqual(sides, [.top, .right, .bottom])
    }

    func testSidesDetectedWhenLargerCollidesOnRight() {
        let entity1 = Entity(
            species: .agent,
            id: "entity1",
            frame: CGRect(x: 0, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let entity2 = Entity(
            species: .agent,
            id: "entity2",
            frame: CGRect(x: 0.9, y: -1, width: 1, height: 3),
            in: testEnv
        )
        let collision = entity1.collision(with: entity2)
        let sides = collision?.sides() ?? []
        XCTAssertEqual(sides, [.top, .right, .bottom])
    }

    func testSidesDetectedWhenObjectCollidesOnTopRight() {
        let entity1 = Entity(
            species: .agent,
            id: "entity1",
            frame: CGRect(x: 0, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let entity2 = Entity(
            species: .agent,
            id: "entity2",
            frame: CGRect(x: 0.9, y: 0.9, width: 1, height: 1),
            in: testEnv
        )
        let collision = entity1.collision(with: entity2)
        let sides = collision?.sides() ?? []
        XCTAssertEqual(sides, [.top, .right])
    }

    func testSidesDetectedWhenObjectCollidesOnBottomRight() {
        let entity1 = Entity(
            species: .agent,
            id: "entity1",
            frame: CGRect(x: 0, y: 0, width: 1, height: 1),
            in: testEnv
        )
        let entity2 = Entity(
            species: .agent,
            id: "entity2",
            frame: CGRect(x: 0.9, y: -0.9, width: 1, height: 1),
            in: testEnv
        )
        let collision = entity1.collision(with: entity2)
        let sides = collision?.sides() ?? []
        XCTAssertEqual(sides, [.right, .bottom])
    }
}

File: Sources/windows/yage/tests/collisions_tests.py
import unittest
from yage.models.collisions import CollisionSide
from yage.models.entity import Entity
from yage.utils.geometry import Rect
from yage.models.species import SPECIES_AGENT
from yage.models.world import World


class CollisionsTests(unittest.TestCase):
    def setUp(self):
        self.test_bounds = Rect(0, 0, 100, 100)
        self.test_env = World("test", self.test_bounds)

    def test_multiple_collisions_can_be_detected(self):
        main = Entity(
            SPECIES_AGENT,
            "main",
            Rect(0, 0, 1, 1),
            self.test_env
        )

        def test_entity(index):
            return Entity(
                SPECIES_AGENT,
                f'other-{index}',
                Rect(0.1 * index, 0, 1, 1),
                self.test_env
            )
        others = [test_entity(i) for i in range(12)]
        collisions = main.collisions(others)
        self.assertEqual(len(collisions), 11)

        more_collisions = main.collisions([main] + others)
        self.assertEqual(len(more_collisions), 11)

    def test_distant_entities_do_not_collide(self):
        entity1 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0, 0, 1, 1),
            self.test_env
        )
        entity2 = Entity(
            SPECIES_AGENT,
            "entity2",
            Rect(2, 2, 1, 1),
            self.test_env
        )
        collision = entity1.collision(entity2)
        self.assertIsNone(collision)

    def test_equal_entities_do_collide(self):
        entity = Entity(SPECIES_AGENT, "someEntity",
                        Rect(0, 0, 0, 0), self.test_env)
        collision = entity.collision(entity)
        self.assertIsNotNone(collision)

    def test_entities_with_same_frame_collide(self):

        entity1 = Entity(SPECIES_AGENT, "entity1",
                         Rect(0, 0, 0, 0), self.test_env)
        entity2 = Entity(SPECIES_AGENT, "entity2",
                         Rect(0, 0, 0, 0), self.test_env)
        collision = entity1.collision(entity2)
        self.assertIsNotNone(collision)

    def test_entities_sharing_one_corner_collide(self):
        entity1 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0, 0, 1, 1),
            self.test_env
        )
        entity2 = Entity(
            SPECIES_AGENT,
            "entity2",
            Rect(1, 1, 1, 1),
            self.test_env
        )
        collision = entity1.collision(entity2)
        expected_intersection = Rect(1, 1, 0, 0)
        self.assertIsNotNone(collision)
        self.assertEqual(collision.intersection, expected_intersection)

    def test_entities_sharing_one_side_collide(self):
        entity1 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0, 0, 1, 1),
            self.test_env
        )
        entity2 = Entity(
            SPECIES_AGENT,
            "entity2",
            Rect(1, 0, 1, 1),
            self.test_env
        )
        collision = entity1.collision(entity2)
        expected_intersection = Rect(1, 0, 0, 1)
        self.assertIsNotNone(collision)
        self.assertEqual(collision.intersection, expected_intersection)

    def test_overlapping_entities_collide(self):
        entity1 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0, 0, 2, 2),
            self.test_env
        )
        entity2 = Entity(
            SPECIES_AGENT,
            "entity2",
            Rect(1, 1, 2, 2),
            self.test_env
        )
        collision = entity1.collision(entity2)
        expected_intersection = Rect(1, 1, 1, 1)
        self.assertIsNotNone(collision)
        self.assertEqual(collision.intersection, expected_intersection)

    def test_overlapping_entities_collide(self):
        entity1 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0, 0, 2, 2),
            self.test_env
        )
        entity2 = Entity(
            SPECIES_AGENT,
            "entity2",
            Rect(1, 1, 2, 2),
            self.test_env
        )
        collision = entity1.collision(entity2)
        expected_intersection = Rect(1, 1, 1, 1)
        self.assertIsNotNone(collision)
        self.assertEqual(collision.intersection, expected_intersection)

    def test_collision_with_ephemeral_entities_are_properly_marked(self):
        entity1 = Entity(SPECIES_AGENT, "entity1", Rect.zero(), self.test_env)
        entity2 = Entity(SPECIES_AGENT, "entity2", Rect.zero(), self.test_env)
        entity2.is_ephemeral = True
        collision = entity1.collision(entity2)
        self.assertTrue(collision.is_ephemeral)

    def test_sides_detected_when_same_height_object_perfectly_collides_on_right(self):
        entity1 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0, 0, 1, 1),
            self.test_env
        )
        entity2 = Entity(
            SPECIES_AGENT,
            "entity2",
            Rect(0.9, 0, 1, 1),
            self.test_env
        )
        collision = entity1.collision(entity2)
        sides = collision.sides()
        self.assertEqual(
            sides, [CollisionSide.TOP, CollisionSide.RIGHT, CollisionSide.BOTTOM])

    def test_sides_detected_when_smaller_collides_on_right(self):
        entity1 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0, 0, 1, 1),
            self.test_env
        )
        entity2 = Entity(
            SPECIES_AGENT,
            "entity2",
            Rect(0.9, 0.25, 1, 0.5),
            self.test_env
        )
        collision = entity1.collision(entity2)
        sides = collision.sides()
        self.assertEqual(
            sides, [CollisionSide.TOP, CollisionSide.RIGHT, CollisionSide.BOTTOM])

    def test_sides_detected_when_larger_collides_on_right(self):
        entity1 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0, 0, 1, 1),
            self.test_env
        )
        entity2 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0.9, -1, 1, 3),
            self.test_env
        )
        collision = entity1.collision(entity2)
        sides = collision.sides()
        self.assertEqual(
            sides, [CollisionSide.TOP, CollisionSide.RIGHT, CollisionSide.BOTTOM])

    def test_sides_detected_when_object_collides_on_top_right(self):
        entity1 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0, 0, 1, 1),
            self.test_env
        )
        entity2 = Entity(
            SPECIES_AGENT,
            "entity2",
            Rect(0.9, 0.9, 1, 1),
            self.test_env
        )
        collision = entity1.collision(entity2)
        sides = collision.sides()
        self.assertEqual(sides, [CollisionSide.TOP, CollisionSide.RIGHT])

    def test_sides_detected_when_object_collides_on_bottom_right(self):
        entity1 = Entity(
            SPECIES_AGENT,
            "entity1",
            Rect(0, 0, 1, 1),
            self.test_env
        )
        entity2 = Entity(
            SPECIES_AGENT,
            "entity2",
            Rect(0.9, -0.9, 1, 1),
            self.test_env
        )
        collision = entity1.collision(entity2)
        sides = collision.sides()
        self.assertEqual(sides, [CollisionSide.RIGHT, CollisionSide.BOTTOM])
